nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/eventhub/GetAzureEventHub.java

```diff
@@ -312,11 +317,13 @@ public void onScheduled(final ProcessContext context) throws ProcessException, U
             receiverFetchTimeout = null;
         }
 
-        final String connectionString = new ConnectionStringBuilder(new URI("amqps://"+namespace+serviceBusEndpoint), eventHubName, policyName, policyKey).toString();
-        setupReceiver(connectionString);
+        final int numThreads = context.getMaxConcurrentTasks();
+        executor = Executors.newScheduledThreadPool(numThreads);
```

Comment:
We should avoid setting up an executor here at all if we can.  NiFi processor API fully supports having multiple tasks/threads for a given processor and already manages the lifecycle correctly and has its own thread pool users can set important configurations of.  Further, we should not do such things in onScheduled.  If such a resource must be created it should be done so lazily on demand on the onTrigger method as the favored pattern.

Reply:
> @mysunnytime can you please describe what testing/verification has been done for this?

Sure.

There are three unit test for the three processors (PutAzureEventHub, GetAzureEventHub, and ConsumeAzureEventHut). They all passes. For this pr, we just upgraded the Event Hub api version and did not change the functionality, so no new test added.

We also test the three components from the Nifi GUI to send/receive events to event hub. Before our updates, processors initialization will fail with ServiceBusException: TransportException (because it uses a really old version of Event Hubs Java client, which uses a really old version of Proton-J, which exhibits this problems when used with current versions of the Java runtime). After the api upgrade, the three components work as expected. Send/receive was successful. ApacheNiFi-eventhub version is also successfully send for us to better understand our customer and provide support.

Reply:
We have to use executor because in the new api we need ScheduledExecutorService to create EventHubClient. But I'm not sure whether we should create the ScheduledExecutorService in onScheduled or onTrigger. 

My understanding is that in **onScheduled**, where we prepare the processor to be ready to work, we set up EventHubClient here, and now we need to create this EventHubClient with a ScheduledExecutorService, and in **onTrigger**, we use EventHubClient to receive msg or send msg. So before we set up EventHubClient, we need to have a ScheduledExecutorService. So it should be in the same (or earlier) place. Also we don't want the ScheduledExecutorService to be created everytime when we wanna send/receive msg (i.e., in onTrigger), so I think it should still be in onScheduled. 

Please tell me what you think, and correct me if I understand onSchedule and onTrigger incorrectly.

Reply:
@mysunnytime it is fine to use the onScheduled to create the necessary connections to Azure event hubs.  For pushing to event hubs it is probably better to use lazy instantiation but in the consume case eager makes more sense and the lifecycle hook of onScheduled is fine.

As far as NiFi and thread pooling yes NiFi has a thread pool.  Basically the whole flow controller is a thread pool.  It isn't generally desired for us to create threads via a new thread pool and give it to some library such as the event hubs sdk.  Rather we'd like to bring a thread to the library to do some meaningful work (fetch data) then process it and repeat.  That allows us to have control over the threads, share them as necessary for splitting work, etc.. while allowing the user to control overall thread count in the system.

Are you certain the only API for leveraging the event hubs SDK is one where we have to hand it over a pool of threads it gets to own then we offer event handlers as data arrives?  How would back-pressure work in that case?  We dont want to keep getting data when we're not ready to process it...  We much prefer a model where we can use our thread to go grab data available and ready and then go do other work and come back.

Reply:
Regarding the thread pool: AMQP is not a request-response protocol. Any AMQP implementation requires the ability to process and generate network activity asynchronously from anything else going on in the application, so it can't just execute on the thread that calls send or receive. Executing on a thread pool supplied by the application is how we allow the application to have control over the thread count.

As for back pressure, that's pretty simple. There is a local buffer (the "prefetch" buffer) of a user-controlled size, and the asynchronous side of the client places messages coming down from the service into that buffer, using AMQP's flow-control mechanisms to cut off the flow when the buffer is full. As the application consumes messages out of the prefetch buffer, the async side uses those flow-control mechanisms to indicate to the service that additional capacity has become available.

Reply:
That context for Azure Event Hubs is helpful.  We still would generally not want to hand off a thread pool to another library unless absolutely necessary.  As we have threads available we would want to offer them to do necessary 'Event Hub interaction' and 'process received data actions' all under our control.  Based on your response though it sounds like the provided threadpool approach is the API offered by event hubs at this point.  OK.

@mysunnytime 

1. Is the threadpool that is created being properly accounted for with the lifecycle of the processor? If the processors is stopped is the thread pool shut down?
2. The number of threads matching the number of tasks can match the number of tasks but this i